bindings/pyroot_experimental/PyROOT/test/rvec_asrvec.py

```diff
@@ -0,0 +1,102 @@
+import unittest
+import ROOT
+import numpy as np
+import sys
+
+
+def get_maximum_for_dtype(dtype):
+    if np.issubdtype(dtype, np.integer):
+        return np.iinfo(dtype).max
+    if np.issubdtype(dtype, np.floating):
+        return np.finfo(dtype).max
+
+def get_minimum_for_dtype(dtype):
+    if np.issubdtype(dtype, np.integer):
+        return np.iinfo(dtype).min
+    if np.issubdtype(dtype, np.floating):
+        return np.finfo(dtype).min
+
+
+class AsRVec(unittest.TestCase):
+    """
+    Tests for the AsRVec feature enabling to adopt memory of Python objects
+    with an array interface member using RVec as C++ container.
+    """
+
+    # Helpers
+    dtypes = [
+        "int32", "int64", "uint32", "uint64", "float32", "float64"
+    ]
+
+    def check_memory_adoption(self, root_obj, np_obj):
+        np_obj[0] = get_maximum_for_dtype(np_obj.dtype)
+        np_obj[1] = get_minimum_for_dtype(np_obj.dtype)
+        self.assertEqual(root_obj[0], np_obj[0])
+        self.assertEqual(root_obj[1], np_obj[1])
+
+    def check_size(self, expected_size, obj):
+        self.assertEqual(expected_size, obj.size())
+
+    # Tests
+    def test_dtypes(self):
+        """
+        Test adoption of numpy arrays with different data types
+        """
+        for dtype in self.dtypes:
+            np_obj = np.empty(2, dtype=dtype)
+            root_obj = ROOT.AsRVec(np_obj)
+            self.check_memory_adoption(root_obj, np_obj)
+            self.check_size(2, root_obj)
+
+    def test_multidim(self):
+        """
+        Test adoption of multi-dimensional numpy arrays
+        """
+        np_obj = np.array([[1, 2], [3, 4]], dtype="float32")
+        rvec = ROOT.AsRVec(np_obj)
+        self.assertEqual(rvec.size(), 4)
+
+    def test_size_zero(self):
+        """
+        Test adoption of numpy array with size 0
+        """
+        np_obj = np.array([], dtype="float32")
+        rvec = ROOT.AsRVec(np_obj)
+        self.assertEqual(rvec.size(), 0)
+
+    def test_adopt_rvec(self):
+        """
+        Test adoption of RVecs
+        """
+        rvec = ROOT.ROOT.VecOps.RVec("float")(1)
+        rvec[0] = 42
+        rvec2 = ROOT.AsRVec(rvec)
+        self.assertEqual(rvec.size(), rvec2.size())
+        self.assertEqual(rvec[0], rvec2[0])
+        rvec2[0] = 43
+        self.assertEqual(rvec[0], rvec2[0])
+
+    def test_ownership(self):
+        """
+        Test ownership of returned RVec (to be owned by Python)
+        """
+        np_obj = np.array([1, 2])
+        rvec = ROOT.AsRVec(np_obj)
+        self.assertEqual(rvec.__python_owns__, True)
+
+    def test_refcount(self):
```

Comment:
This is tricky. We do not increase the refcount of the numpy array we use for adoption. However, now it could happen that the GC destroys the memory of the numpy array and our adopted rvec points to invalid memory.
Still, I think this is the way to go.

@etejedor Do you agree?

Reply:
What is the disadvantage of increasing the reference of the numpy array and prevent any premature deletion?
How do other libraries that adopt numpy arrays handle this case?

Reply:
Hm good questions, just checked:

```python
import ROOT
import numpy as np
import sys

rvec = ROOT.ROOT.VecOps.RVec("float")(1)
print(sys.getrefcount(rvec))
# Returns: 2

npy = np.asarray(rvec)
print(sys.getrefcount(rvec))
# Returns: 3

del npy
print(sys.getrefcount(rvec))
# Returns: 2
```

Reply:
So yes, they increment the refcount. But who decreases the refcount after the adopted container is deleted? How does the original container know about this?

Reply:
I see only the option to add the adopted object to the dictionary of the rvec. i'll give it a try.

Reply:
ok, i've pushed a fix. now the returned `RVec` has the adopted object as the attribute `__adopted__` (see the tests). Now we have this behaviour:

```python
import ROOT
import numpy as np
import sys

npy = np.array([1, 2])
print(sys.getrefcount(npy))
# Returns: 2

rvec = ROOT.AsRVec(npy)
rvec.__adopted__ = npy
print(sys.getrefcount(npy))
# Returns: 3

del rvec
print(sys.getrefcount(npy))
# Returns: 2
```